reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,948 @@
+const assert = require('assert');
+import { CreateIterResultObject, InvokeOrNoop, PromiseInvokeOrNoop, typeIsObject } from './helpers';
+
+export default class ReadableByteStream {
+  constructor(underlyingByteSource = {}) {
+    this._state = 'readable';
+    this._reader = undefined;
+    this._storedError = undefined;
+
+    this._controller = undefined;
+    this._controller = new ReadableByteStreamController(this, underlyingByteSource);
+  }
+
+  cacnel(reason) {
+    if (IsReadableByteStream(this) === false) {
+      return Promise.reject(
+          new TypeError('ReadableByteStream.prototype.cancel can only be used on a ReadableByteStream'));
+    }
+
+    if (IsReadableByteStreamLocked(this)) {
+      return Promise.reject(new TypeError('Cannot cancel a stream that already has a reader'));
+    }
+
+    return CancelReadableByteStream(this, reason);
+  }
+
+  getByobReader() {
+    if (IsReadableByteStream(this) === false) {
+      throw new TypeError('ReadableByteStream.prototype.getByobReader can only be used on a ReadableByteStream');
+    }
+
+    return new ReadableByteStreamByobReader(this);
+  }
+
+  getReader() {
+    if (IsReadableByteStream(this) === false) {
+      throw new TypeError('ReadableByteStream.prototype.getReader can only be used on a ReadableByteStream');
+    }
+
+    return new ReadableByteStreamReader(this);
+  }
+}
+
+class ReadableByteStreamController {
+  constructor(controlledReadableByteStream, underlyingByteSource) {
+    if (IsReadableByteStream(controlledReadableByteStream) === false) {
+      throw new TypeError('ReadableByteStreamController can only be constructed with a ReadableByteStream instance');
+    }
+
+    if (controlledReadableByteStream._controller !== undefined) {
+      throw new TypeError(
+          'ReadableByteStreamController instances can only be created by the ReadableByteStream constructor');
+    }
+
+    const pullFunction = underlyingByteSource['pull'];
```

Comment:
I prefer letting these fail at call time instead of immediately. The values could change over time and this kind of aggressive type checking is unnecessary.